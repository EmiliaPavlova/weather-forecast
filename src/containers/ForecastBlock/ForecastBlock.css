.forecast_wrapper {
  display: flex;
  justify-content: center;
  min-width: 1000px;
  margin: 20px 50px;
}

.day {
  flex-basis: 20%;
  max-width: 200px;
  padding: 10px;
  border-radius: 14px;
  box-shadow: inset 0px 0px 5px 0px #a7a7a7;
  cursor: pointer;
}

.selected_day {
  box-shadow: inset 0px 0px 5px 0px #000000;
}

.time {
  flex-basis: 12.5%;
  max-width: 125px;
  padding: 10px;
  border-radius: 4px;
  box-shadow: inset 0px 0px 3px 0px #a7a7a7;
}

.forecast_label {
  margin-left: 50px;
  margin-top: 40px;
}