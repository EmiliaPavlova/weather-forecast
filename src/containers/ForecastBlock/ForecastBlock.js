import React, { useState, useMemo } from 'react';
import moment from 'moment';
import CardDay from '../../components/CardDay/CardDay';
import CardTime from '../../components/CardTime/CardTime';
import './ForecastBlock.css';

const daysToDisplay = 5;

export default function ForecastBlock({ forecast }) {
  const [selectedDay, setSelectedDay] = useState();

  const mapDayData = (day) => {
    return {
      day: moment.unix(day.dt).format('dddd'),
      date: moment.unix(day.dt).format('DD MMMM'),
      time: moment.unix(day.dt).format('HH'),
      weather: day.weather[0].main,
      description: day.weather[0].description,
      icon: day.weather[0].icon,
      temp: day.main.temp,
      tempMin: day.main.temp_min,
      tempMax: day.main.temp_max,
      feelsLike: day.main.feels_like,
    }
  }

  const forecastByDays = useMemo(() => {
    let days = [];
    forecast && forecast.map(day => {
      let date = moment.unix(day.dt).format('DD');
      return days[date] = days[date]
        ? [...days[date], mapDayData(day)]
        : [mapDayData(day)];
    })
    days = days.filter(el => !!el);
    if (Object.keys(days).length > daysToDisplay) {
      days.splice(daysToDisplay)
    }
    return days;
  }, [forecast]);

  const getProp = (array, prop) => array.find(slot => slot.time === '12')
    ? array.find(slot => slot.time === '12')[prop]
    : array[0][prop];

  const getTempMin = (array) => JSON.stringify(array
    .reduce((prev, current) => (prev.tempMin < current.tempMin) ? prev : current ));

  const getTempMax = (array) => JSON.stringify(array
    .reduce((prev, current) => (prev.tempMax > current.tempMax) ? prev : current ));

  return (
    <>
      <div className='forecast_wrapper'>
        {forecastByDays.map((day, index) => {
          return (
            <div className={`day${day === selectedDay ? ' selected_day' : ''}`} key={index} onClick={() => setSelectedDay(day)}>
              <CardDay
                day={day[0].day}
                date={day[0].date}
                weather={getProp(day, 'weather')}
                description={getProp(day, 'description')}
                icon={getProp(day, 'icon')}
                temp={getProp(day, 'temp')}
                feelsLike={getProp(day, 'feelsLike')}
                tempMin={JSON.parse(getTempMin(day)).tempMin}
                tempMax={JSON.parse(getTempMax(day)).tempMax}
              />
            </div>
          )
        })}
      </div>
      <div className='forecast_label'>{selectedDay ? `Forecast by hours for ${selectedDay[0].day}, ${selectedDay[0].date}:` : 'Click on a day to see a detailed forecast.'}</div>
      <div className='forecast_wrapper'>
        {selectedDay && selectedDay.map((day, index) => {
          return (
            <div className='time' key={index}>
              <CardTime
                time={day.time}
                icon={day.icon}
                description={day.description}
                temp={day.temp}
                feelsLike={day.feelsLike}
              />
            </div>
          )
        })}
      </div>
    </>
  )
};