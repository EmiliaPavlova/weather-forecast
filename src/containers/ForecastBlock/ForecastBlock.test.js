import React from 'react';
import { screen, cleanup, fireEvent } from '@testing-library/react';
import { shallow } from 'enzyme';
import ForecastBlock from './ForecastBlock';

const mockForecast = [
  {
    dt: 1650067200,
    weather: [{ main: 'Clouds', description: 'scattered clouds', icon: '03n'}],
    main: [{ temp: 284.2, feels_like: 282.98, temp_min: 283.57, temp_max: 284.2 }]
  }, {
    dt: 1650078000,
    weather: [{ main: 'Clouds', description: 'scattered clouds', icon: '03n'}],
    main: [{ temp: 283.24, feels_like: 282.11, temp_min: 282.6, temp_max: 283.24 }]
  }, {
    dt: 1650088800,
    weather: [{ main: 'Clouds', description: 'broken clouds', icon: '04n'}],
    main: [{ temp: 286.01, feels_like: 284.79, temp_min: 286.01, temp_max: 286.01 }]
  },

  {
    dt: 1650099600,
    weather: [{ main: 'Clouds', description: 'overcast clouds', icon: '04n'}],
    main: [{ temp: 290, feels_like: 289.05, temp_min: 290, temp_max: 2901 }]
  }
];

describe('ForecastBlock component', () => {
  afterEach(cleanup);

  it('matches the snapshot', () => {
    const wrapper = shallow(<ForecastBlock forecast={mockForecast} />);
    expect(wrapper).toMatchSnapshot();
  });
})