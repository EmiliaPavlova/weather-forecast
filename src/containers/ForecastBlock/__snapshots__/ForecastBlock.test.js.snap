// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastBlock component matches the snapshot 1`] = `
<Fragment>
  <div
    className="forecast_wrapper"
  >
    <div
      className="day"
      key="0"
      onClick={[Function]}
    >
      <CardDay
        date="16 April"
        day="Saturday"
        description="overcast clouds"
        icon="04n"
        weather="Clouds"
      />
    </div>
  </div>
  <div
    className="forecast_label"
  >
    Click on a day to see a detailed forecast.
  </div>
  <div
    className="forecast_wrapper"
  />
</Fragment>
`;
